Encryption.md

```diff
@@ -0,0 +1,221 @@
+
+### Encrypted footer mode
+
+In files with sensitive column data, a good security practice is to encrypt not only the 
+secret columns, but also the file footer metadata, with a separate footer key. This hides
+the file schema / column names, number of rows, key-value properties, column sort order, 
+column data offset and size, list of encrypted columns and metadata of the column encryption keys. 
+It also makes the footer tamper-proof.
+
+The columns encrypted with the same key as the footer, don't split the ColumnMetaData from the 
+ColumnChunks, leaving it at the original location, `optional ColumnMetaData meta_data`. This field
+is not set for columns enrypted with a column-specific key.
+
+A Thrift-serialized `FileCryptoMetaData` structure is written after the footer. It contains 
+information on the file encryption algorithm and on the footer (offset in 
+the file, and an optional key metadata, with a maximal length of 512). Then 
+the length of this structure is written, as a 4-byte little endian integer. Then a final 
+magic string, "PARE".
+
+Only the `FileCryptoMetaData` is written as a plaintext, all other file parts are protected
+(as needed) with appropriate keys.
+
+### Plaintext footer mode
+
+This mode allows older Parquet versions (released before the encryption support) to access unencrypted 
+columns in encrypted files - at a price of leaving certain metadata fields unprotected in these files 
+(not encrypted or tamper-proofed). The plaintext footer mode can be useful during a transitional period 
+in organizations 
+where some frameworks can't be upgraded to a new Parquet library for a while. Data writers will
+upgrade and run with a new Parquet version, producing encrypted files in this mode. Data readers, 
+working with a sensitive data, will also upgrade to a new Parquet library. But other readers that
+don't need the sensitive columns, can continue working with an older Parquet version. They will be 
+able to access plaintext columns in encrypted files. An older reader, trying to access a sensitive 
+column data in a ".parquet.encrypted" file with a plaintext footer, will get an  exception. More
+specifically, a Thrift parsing exception on an encrypted `PageHeader` structure. Again, using older
+Parquet readers for encrypted files is a temporary solution.
+
+In the plaintext footer mode, the `optional ColumnMetaData meta_data` is set in the `ColumnChunk` 
+structure for all columns, but is stripped of the statistics for the sensitive (encrypted) columns. 
+These statistics are available for new readers with the column key - they fetch the split ColumnMetaData, 
+and decrypt it to get all metadata values.
+
+An `encryption_algorithm` field is set at the FileMetaData structure. Then the footer is written as usual, 
+followed by its length (4-byte little endian integer) and a final magic string, "PAR1".
+
+
+
+### New fields for vectorized readers
```

Comment:
Could you please elaborate a bit more on how this currently works without encryption, why can it not work the same way with encryption and how the new fields solve this?

Reply:
Sure, will do.